README.md
# Reloader

More control over hot code push reloading for your mobile apps. A replacement
for [`mdg:reload-on-resume`](https://github.com/meteor/mobile-packages/blob/master/packages/mdg:reload-on-resume/README.md)
with more options and better UX.

Before using this package we recommend that you understand what Hot Code Push is, you can learn all about it [here](https://guide.meteor.com/hot-code-push.html)

We provide two ways for you to handle your app code updates:

- Always reload
- Reload when the app allows (recommended)

### Always reload

You don't need to configure anything, your app is going to reload as soon as the
code is received.

If you want you can inform `launchScreenDelay` (0 by default) in milliseconds to
hold your splashscreen longer, avoiding a flash when the app is starting and
reloading.

```json
"public": {
  "packages": {
    "jaskinn:reloader": {
      "launchScreenDelay": 200
    }
  }
}
```

### Reload when the app allows

We recommend this method as with it you can control when your app is going to
reload. You can even delegate this decision to the final user.

In this case you must use `automaticInitialization` as `false` in your settings.

```json
"public": {
  "packages": {
    "jaskinn:reloader": {
      "automaticInitialization": false
    }
  }
}
```

You also need to call 
`Reloader.initialize` in the render or initialization of your app providing a function (can be async) in the property `beforeReload`.

## Installing

```sh
meteor add jaskinn:reloader
meteor remove mdg:reload-on-resume
```

## Configuration Options

### idleCutoff

Default: `1000 * 60 * 5 // 5 minutes`

How long (in ms) can an app be idle before we consider it a start and not a
resume. Applies only when `check: 'everyStart'`. Set to `0` to never check on
resume.

### launchScreenDelay

Default: `0`

How long the splash screen will be visible. It's useful to avoid your app being rendered just for a few milliseconds and then refreshing.

### automaticInitialization

Default: `true`

If you want to initialize the `reloader` yourself you need to turn
off `automaticInitialization`. This is useful when you want to provide code to
some callback as this is not possible using JSON initialization.

You can provide your callbacks calling Reloader.initialize(), for example:

```js
ReloaderCordova.initialize({
  beforeReload(ok, nok) {
    const isOkToReload = confirm('Your app will load now, ok?');
    if (isOkToReload) {
      ok();
      return;
    }
    nok();
  },
});
```

## Example with React

File: `Routes.js` (where we render the routes)
```javascript

export const Routes = () => {
  useEffect(() => initializeReloader(), []);

  return (
    <Switch>
      // React router routes...
    </Switch>
  );
}
```

File: `initializeReloader.js`
```javascript
import { Reloader } from 'meteor/quave:reloader';
import { loggerClient } from 'meteor/quave:logs/loggerClient';
import { showConfirm } from './ConfirmationDialog';
import { methodCall } from '../../methods/methodCall';
import { version } from '../../version';

export const initializeReloader = () => {
  loggerClient.info({ message: 'initializeReloader' });
  Reloader.initialize({
    async beforeReload(updateApp, holdAppUpdate) {
      loggerClient.info({ message: 'initializeReloader beforeReload' });
      let appUpdateData = {};
      try {
        appUpdateData =
          (await methodCall('getAppUpdateData', { clientVersion: version })) ||
          {};
      } catch (e) {
        loggerClient.info({
          message: 'forcing app reload because getAppUpdateData is breaking',
        });
        updateApp();
        return;
      }
      loggerClient.info({
        message: 'initializeReloader beforeReload appUpdateData',
        appUpdateData,
      });
      if (appUpdateData.ignore) {
        loggerClient.info({
          message:
            'initializeReloader beforeReload appUpdateData ignore is true',
          appUpdateData,
        });
        return;
      }
      const cancelAction = appUpdateData.forceUpdate
        ? updateApp
        : holdAppUpdate;
      try {
        const message = appUpdateData.forceUpdate
          ? 'Precisamos atualizar o aplicativo. É rapidinho!'
          : 'Deseja atualizar agora? É rapidinho!';
        const result = await showConfirm({
          autoFocus: false,
          title: appUpdateData.title || 'Atualização disponível',
          content: appUpdateData.message || message,
          confirmText: appUpdateData.actionLabel || 'Beleza',
          cancelText: appUpdateData.noActionLabel || 'Mais tarde',
          hideCancel: !!appUpdateData.forceUpdate,
          dismiss: cancelAction,
          onCancel() {
            loggerClient.info({
              message: 'initializeReloader beforeReload onCancel',
              appUpdateData,
            });
            cancelAction();
          },
        });
        loggerClient.info({
          message: `initializeReloader beforeReload showConfirm result is ${result}`,
          appUpdateData,
        });
        if (result) {
          loggerClient.info({
            message: 'initializeReloader beforeReload showConfirm ok',
            appUpdateData,
          });
          updateApp();
          return;
        }
        loggerClient.info({
          message: 'initializeReloader beforeReload showConfirm nok',
          appUpdateData,
        });
        cancelAction();
      } catch (e) {
        loggerClient.info({
          message: 'initializeReloader beforeReload showConfirm catch call nok',
          appUpdateData,
        });
        cancelAction();
      }
    },
  });
};

```

File: `getAppUpdateData.js`
```javascript
import { Meteor } from 'meteor/meteor';
import { logger } from 'meteor/quave:logs/logger';
import { AppUpdatesCollection } from '../db/AppUpdatesCollection';
import { version } from '../version';

Meteor.methods({
  getAppUpdateData({ clientVersion } = {}) {
    this.unblock();

    if (Meteor.isClient) return null;

    const appUpdate = AppUpdatesCollection.findOne() || {};

    const result = {
      ...appUpdate,
      ...(appUpdate.ignoreVersions &&
      appUpdate.ignoreVersions.length &&
      appUpdate.ignoreVersions.includes(version)
        ? { ignore: true }
        : {}),
      version,
    };
    logger.info({
      message: `getAppUpdateData clientVersion=${clientVersion}, newClientVersion=${version}, ${JSON.stringify(
        result
      )}`,
      appUpdateData: appUpdate,
      appUpdateResult: result,
      clientVersion,
    });
    return result;
  },
});

```


CHANGELOG.md
# Changelog

## 2.0.3 - 2021-02-10
### Fixed

- Fix for intermittent "undefined is not an object (evaluating 'navigator.splashscreen.show')" error on resume in iOS.
 
## 2.0.2 - 2020-10-19
### New feature
You can control from your app code when your app should update.

### Clean up
We removed options that we don't believe are necessary, read more details [here](./README.md)

## 1.6.0 - 2020-06-12
### Config using Meteor.settings
- Config now is set using `Meteor.settings.public.packages.reloader` object

## past (before quave fork)
## [1.5.0] - 2019-02-08
### Config using Meteor.settings
- Config now is set using `Meteor.settings.public.reloader` object
- Added more logs if `Meteor.settings.public.reloader.debug` is true

## [1.4.0] - 2018-10-03
### Independency from Blaze
- Removed Blaze dependency
- Removed underscore dependency

## [1.3.0] - 2017-11-13
### Updated

- `cordova-plugin-splashscreen` from `4.0.0` to `4.1.0`

## [1.2.2] - 2016-10-05
### Fixed

- Fix reloading when url contains a hash ([#10](https://github.com/jamielob/reloader/issues/10))

## [1.2.0] - 2016-04-26
### Added
- `launchScreenDelay` option
- tests


reloader-tests.js
// NOTE: I think our Reload._onMigrate messes with test app reloading?
// See "Cannot read property '0' of undefined" in browser console
// after editing the code
//
// In order to get test output to update, you may need to
// click the "refresh" header

// http://chaijs.com/api/assert/
import { assert } from "meteor/practicalmeteor:chai";
import { Reloader } from "reloader";

// http://sinonjs.org/
import sinon from "sinon";

describe("refresh", function() {
  beforeEach(function() {
    Reloader.reload = sinon.spy(); // todo save original and repair if needed later
  });

  describe("startAndResume", function() {
    before(function() {
      Reloader.configure({
        refresh: "startAndResume"
      });
    });

    it("reloads on resume when update is available", function() {
      Reloader.updateAvailable.set(true);

      Reloader._onResume();
      assert.isTrue(Reloader.reload.called);
    });

    it("doesn't reload on resume when update isn't available", function() {
      Reloader.updateAvailable.set(false);

      Reloader._onResume();

      assert.isFalse(Reloader.reload.called);
    });
  });

  describe("start", function() {
    before(function() {
      Reloader.configure({
        refresh: "start"
      });
    });

    it("doesn't reload on resume when update is available", function() {
      Reloader.updateAvailable.set(true);

      Reloader._onResume();

      assert.isFalse(Reloader.reload.called);
    });
  });
});

describe("check", function() {
  // should call / not call _checkForUpdate
});


reloader-cordova.js
import {Meteor} from 'meteor/meteor';
import {Tracker} from 'meteor/tracker';
import {ReactiveVar} from 'meteor/reactive-var';
import {LaunchScreen} from 'meteor/launch-screen';
import {settings, debugFn, PACKAGE_NAME} from './common';

const DEFAULT_OPTIONS = {
  idleCutoff: 1000 * 60 * 5, // 5 minutes
  automaticInitialization: true,
  launchScreenDelay: 0,
};

debugFn('starting - DEFAULT_OPTIONS', {DEFAULT_OPTIONS});
debugFn('starting - settings', {settings});
const options = Object.assign({}, DEFAULT_OPTIONS, settings);

debugFn('starting - options', {options});

const defaultRetry = () => console.log(`[${PACKAGE_NAME}] no retry function yet`);

const launchScreen = LaunchScreen.hold();

let initialized = false;

const Reloader = {
  _options: {},
  // eslint-disable-next-line no-console
  _retry: defaultRetry,
  updateAvailable: new ReactiveVar(false),
  isChecked: new ReactiveVar(false),

  debug(message, context) {
    debugFn(message, {
      ...context,
      updateAvailable: this.updateAvailable.get(),
      isChecked: this.isChecked.get(),
      _options: this._options,
    });
  },

  initialize(optionsParam = {}) {
    if (initialized) {
      return;
    }
    initialized = true;
    this._options = Object.assign({}, options, optionsParam);
    this._onPageLoad();
  },

  prepareToReload() {
    this.debug('prereload - show splashscreen');
    // Show the splashscreen
    navigator.splashscreen.show();

    const currentDate = Date.now();
    this.debug('prereload - reloaderWasRefreshed', {currentDate});
    // Set the refresh flag
    localStorage.setItem('reloaderWasRefreshed', currentDate);
  },

  reloadNow() {
    this.debug('reloadNow');
    if (this._isCheckBeforeReload() && !this.isChecked.get()) {
      this.debug(
        'not reloading because beforeReload is provided and it is not checked yet'
      );
      return;
    }
    this.prepareToReload();

    // We'd like to make the browser reload the page using location.replace()
    // instead of location.reload(), because this avoids validating assets
    // with the server if we still have a valid cached copy. This doesn't work
    // when the location contains a hash however, because that wouldn't reload
    // the page and just scroll to the hash location instead.
    if (window.location.hash || window.location.href.endsWith('#')) {
      this.debug('reloadNow - reload');
      window.location.reload();
    } else {
      this.debug('reloadNow - replace');
      window.location.replace(window.location.href);
    }
  },

  // Should check if a cold start and (either everyStart is set OR firstStart
  // is set and it's our first start)
  _shouldCheckForUpdateOnStart() {
    this.debug('_shouldCheckForUpdateOnStart');

    const isColdStart = !localStorage.getItem('reloaderWasRefreshed');
    const reloaderLastStart = localStorage.getItem('reloaderLastStart');
    this.debug('_shouldCheckForUpdateOnStart - info', {
      isColdStart,
      check: this._options.check,
      reloaderLastStart,
    });
    const should =
      isColdStart;

    this.debug('_shouldCheckForUpdateOnStart - should', {should});
    return should;
  },

  // Check if the idleCutoff is set AND we exceeded the idleCutOff limit AND the everyStart check is set
  _shouldCheckForUpdateOnResume() {
    this.debug('_shouldCheckForUpdateOnResume');
    const reloaderLastPause = localStorage.getItem('reloaderLastPause');
    // In case a pause event was missed, assume it didn't make the cutoff
    if (!reloaderLastPause) {
      this.debug('_shouldCheckForUpdateOnResume no reloaderLastPause');
      return false;
    }

    // Grab the last time we paused
    const lastPause = Number(reloaderLastPause);

    // Calculate the cutoff timestamp
    const idleCutoffAt = Number(Date.now() - this._options.idleCutoff);

    this.debug('_shouldCheckForUpdateOnResume - info', {
      idleCutoff: this._options.idleCutoff,
      check: this._options.check,
      lastPause,
      idleCutoffAt,
    });
    return (
      this._options.idleCutoff &&
      lastPause < idleCutoffAt
    );
  },

  _waitForUpdate(computation) {
    this.debug('_waitForUpdate');
    // Check if we have a HCP after the check timer is up
    Meteor.setTimeout(() => {
      // If there is a new version available
      if (this.updateAvailable.get()) {
        this.debug('_waitForUpdate - reloadNow');
        this.reloadNow();
      } else {
        // Stop waiting for update
        if (computation) {
          computation.stop();
        }

        this.debug('prereload - release launchScreen');
        launchScreen.release();

        if (
          !navigator ||
          !navigator.splashscreen ||
          !navigator.splashscreen.hide
        ) {
          console.warn(
            `[${PACKAGE_NAME}] navigator.splashscreen.hide not available`
          );
          return;
        }
        this.debug('prereload - hide splashscreen');
        navigator.splashscreen.hide();
      }
    }, 0);
  },

  _checkForUpdate() {
    this.debug('_checkForUpdate');
    if (this.updateAvailable.get()) {
      // Check for an even newer update
      this.debug('_checkForUpdate - check for an even newer update');
      this._waitForUpdate();
    } else {
      // Wait until update is available, or give up on timeout
      Tracker.autorun(c => {
        if (this.updateAvailable.get()) {
          this.debug('_checkForUpdate - reloadNow');
          this.reloadNow();
        }

        this._waitForUpdate(c);
      });
    }
  },

  _onPageLoad() {
    this.debug('_onPageLoad');
    if (this._shouldCheckForUpdateOnStart()) {
      this._checkForUpdate();
    } else {
      Meteor.setTimeout(() => {
        this.debug('_onPageLoad - release launchScreen');
        launchScreen.release();

        // Reset the reloaderWasRefreshed flag
        localStorage.removeItem('reloaderWasRefreshed');
      }, this._options.launchScreenDelay); // Short delay helps with white flash
    }
  },

  _onResume() {
    this.debug('_onResume');
    const shouldCheck = this._shouldCheckForUpdateOnResume();

    localStorage.removeItem('reloaderLastPause');

    if (shouldCheck) {
      this.debug('_onResume - show splashscreen');
      navigator.splashscreen.show();

      this._checkForUpdate();
      return;
    }
  },

  _isCheckBeforeReload() {
    this.debug('_isCheckBeforeReload');
    return !!this._options.beforeReload && typeof this._options.beforeReload === 'function';
  },

  _callBeforeLoad() {
    this.debug('_callBeforeLoad');
    const updateApp = () => {
      this.debug('_callBeforeLoad set isChecked to true');
      this.isChecked.set(true);
      this._retry();
    };
    const holdAppUpdate = () => {
      this.debug('_callBeforeLoad set isChecked to false');
      this.isChecked.set(false);
    };
    this._options.beforeReload(updateApp, holdAppUpdate);
  },

  // https://github.com/meteor/meteor/blob/devel/packages/reload/reload.js#L104-L122
  _onMigrate(retry) {
    this.debug('_onMigrate');
    this._retry = retry || this._retry;
    if (!this._isCheckBeforeReload() || this.isChecked.get()) {
      // we are calling prepareToReload because as we are returning true the reload
      // will happen in the reload package then we are updating our timestamps
      this.prepareToReload();

      this.isChecked.set(false);
      return [true, {}];
    }

    if (this._isCheckBeforeReload()) {
      this._callBeforeLoad();
    }

    // Set the flag
    this.updateAvailable.set(true);

    // Don't refresh yet
    return [false];
  },
};

if (options.automaticInitialization) {
  Reloader.initialize();
}

// Set the last start flag
localStorage.setItem('reloaderLastStart', Date.now());

// Watch for the app resuming
document.addEventListener(
  'resume',
  () => {
    Reloader._onResume();
  },
  false
);

localStorage.removeItem('reloaderLastPause');

// Watch for the device pausing
document.addEventListener(
  'pause',
  () => {
    // Save to localStorage
    localStorage.setItem('reloaderLastPause', Date.now());
  },
  false
);

// Capture the reload
// import { Reload } from 'meteor/reload' is not working
// eslint-disable-next-line no-undef
Reload._onMigrate(`${PACKAGE_NAME}`, retry => Reloader._onMigrate(retry));

export {Reloader};


reloader.js
import { Meteor } from 'meteor/meteor';
import { Tracker } from 'meteor/tracker';
import { ReactiveVar } from 'meteor/reactive-var';
import { LaunchScreen } from 'meteor/launch-screen';
import { getSettings } from 'meteor/quave:settings';

export const PACKAGE_NAME = 'quave:reloader';
export const settings = getSettings({ packageName: PACKAGE_NAME });

/**
 *  check: Match.Optional(Match.OneOf('everyStart', 'firstStart', false)),
 *  checkTimer: Match.Optional(Match.Integer),
 *  refresh: Match.Optional(
 *    Match.OneOf('startAndResume', 'start', 'instantly')
 *  ),
 *  idleCutoff: Match.Optional(Match.Integer),
 *  launchScreenDelay: Match.Optional(Match.Integer),
 *  debug: Boolean
 */
const DEFAULT_OPTIONS = {
  check: 'everyStart',
  checkTimer: 0,
  refresh: 'startAndResume',
  idleCutoff: 1000 * 60 * 5, // 5 minutes
  launchScreenDelay: 100,
};

const debug = (message, context) => {
  if (!settings.debug) {
    return;
  }
  // eslint-disable-next-line no-console
  console.log(`[${PACKAGE_NAME}] ${message}`, JSON.stringify(context));
};

const launchScreen = LaunchScreen.hold();

const Reloader = {
  _options: {},
  updateAvailable: new ReactiveVar(false),

  initialize() {
    debug('initialize - DEFAULT_OPTIONS', { DEFAULT_OPTIONS });
    debug('initialize - settings', { settings });
    const optionsWithDefaults = Object.assign({}, DEFAULT_OPTIONS, settings);

    Object.assign(this._options, optionsWithDefaults);
    debug('initialize - options', { _options: this._options });
  },

  prereload() {
    debug('prereload - show splashscreen');
    // Show the splashscreen
    navigator.splashscreen.show();

    const currentDate = Date.now();
    debug('prereload - reloaderWasRefreshed', { currentDate });
    // Set the refresh flag
    localStorage.setItem('reloaderWasRefreshed', currentDate);
  },

  reload() {
    debug('reload');
    this.prereload();

    // We'd like to make the browser reload the page using location.replace()
    // instead of location.reload(), because this avoids validating assets
    // with the server if we still have a valid cached copy. This doesn't work
    // when the location contains a hash however, because that wouldn't reload
    // the page and just scroll to the hash location instead.
    if (window.location.hash || window.location.href.endsWith('#')) {
      debug('reload - reload');
      window.location.reload();
    } else {
      debug('reload - replace');
      window.location.replace(window.location.href);
    }
  },

  // Should check if a cold start and (either everyStart is set OR firstStart
  // is set and it's our first start)
  _shouldCheckForUpdateOnStart() {
    debug('_shouldCheckForUpdateOnStart');
    if (!this._options.check) {
      debug('_shouldCheckForUpdateOnStart - check false', {
        check: this._options.check,
      });
      return false;
    }

    const isColdStart = !localStorage.getItem('reloaderWasRefreshed');
    const reloaderLastStart = localStorage.getItem('reloaderLastStart');
    debug('_shouldCheckForUpdateOnStart - info', {
      isColdStart,
      check: this._options.check,
      reloaderLastStart,
    });
    const should =
      isColdStart &&
      (this._options.check === 'everyStart' ||
        (this._options.check === 'firstStart' && !reloaderLastStart));

    debug('_shouldCheckForUpdateOnStart - should', { should });
    return should;
  },

  // Check if the idleCutoff is set AND we exceeded the idleCutOff limit AND the everyStart check is set
  _shouldCheckForUpdateOnResume() {
    debug('_shouldCheckForUpdateOnResume');
    if (!this._options.check) {
      debug('_shouldCheckForUpdateOnResume - check false', {
        check: this._options.check,
      });
      return false;
    }

    const reloaderLastPause = localStorage.getItem('reloaderLastPause');
    // In case a pause event was missed, assume it didn't make the cutoff
    if (!reloaderLastPause) {
      debug('_shouldCheckForUpdateOnResume no reloaderLastPause');
      return false;
    }

    // Grab the last time we paused
    const lastPause = Number(reloaderLastPause);

    // Calculate the cutoff timestamp
    const idleCutoffAt = Number(Date.now() - this._options.idleCutoff);

    debug('_shouldCheckForUpdateOnResume - info', {
      idleCutoff: this._options.idleCutoff,
      check: this._options.check,
      lastPause,
      idleCutoffAt,
    });
    return (
      this._options.idleCutoff &&
      lastPause < idleCutoffAt &&
      this._options.check === 'everyStart'
    );
  },

  _waitForUpdate(computation) {
    debug('_waitForUpdate');
    // Check if we have a HCP after the check timer is up
    Meteor.setTimeout(() => {
      // If there is a new version available
      if (this.updateAvailable.get()) {
        debug('_waitForUpdate - reload');
        this.reload();
      } else {
        // Stop waiting for update
        if (computation) {
          computation.stop();
        }

        debug('prereload - release launchScreen');
        launchScreen.release();

        if (
          !navigator ||
          !navigator.splashscreen ||
          !navigator.splashscreen.hide
        ) {
          console.warn(
            `[${PACKAGE_NAME}] navigator.splashscreen.hide not available`
          );
          return;
        }
        debug('prereload - hide splashscreen');
        navigator.splashscreen.hide();
      }
    }, this._options.checkTimer || 0);
  },

  _checkForUpdate() {
    debug('_checkForUpdate');
    if (this.updateAvailable.get()) {
      // Check for an even newer update
      debug('_checkForUpdate - check for an even newer update');
      this._waitForUpdate();
    } else {
      // Wait until update is available, or give up on timeout
      Tracker.autorun(c => {
        if (this.updateAvailable.get()) {
          debug('_checkForUpdate - reload');
          this.reload();
        }

        this._waitForUpdate(c);
      });
    }
  },

  _onPageLoad() {
    debug('_onPageLoad');
    if (this._shouldCheckForUpdateOnStart()) {
      this._checkForUpdate();
    } else {
      Meteor.setTimeout(() => {
        debug('_onPageLoad - release launchScreen');
        launchScreen.release();

        // Reset the reloaderWasRefreshed flag
        localStorage.removeItem('reloaderWasRefreshed');
      }, this._options.launchScreenDelay); // Short delay helps with white flash
    }
  },

  _onResume() {
    debug('_onResume');
    const shouldCheck = this._shouldCheckForUpdateOnResume();

    localStorage.removeItem('reloaderLastPause');

    if (shouldCheck) {
      debug('_onResume - show splashscreen');
      navigator.splashscreen.show();

      this._checkForUpdate();
      return;
    }

    // If we don't need to do an additional check
    // Check if there's a new version available already AND we need to refresh on resume
    if (
      this.updateAvailable.get() &&
      this._options.refresh === 'startAndResume'
    ) {
      this.reload();
    }
  },

  // https://github.com/meteor/meteor/blob/devel/packages/reload/reload.js#L104-L122
  _onMigrate() {
    debug('_onMigrate');
    if (this._options.refresh === 'instantly') {
      this.prereload();

      return [true, {}];
    }

    // Set the flag
    this.updateAvailable.set(true);

    // Don't refresh yet
    return [false];
  },
};

Reloader.initialize();
Reloader._onPageLoad();

// Set the last start flag
localStorage.setItem('reloaderLastStart', Date.now());

// Watch for the app resuming
document.addEventListener(
  'resume',
  () => {
    Reloader._onResume();
  },
  false
);

localStorage.removeItem('reloaderLastPause');

// Watch for the device pausing
document.addEventListener(
  'pause',
  () => {
    // Save to localStorage
    localStorage.setItem('reloaderLastPause', Date.now());
  },
  false
);

// Capture the reload
// import { Reload } from 'meteor/reload' is not working
// eslint-disable-next-line no-undef
Reload._onMigrate(`${PACKAGE_NAME}`, retry => Reloader._onMigrate(retry));

export { Reloader };
